glean-core/src/metrics/dual_labeled_counter.rs

```diff
@@ -0,0 +1,375 @@
+                    meta,
+                    ErrorType::InvalidLabel,
+                    format!("Invalid `DualLabeledCounter` label format: {label:?}"),
+                    None,
+                );
+                key = OTHER_LABEL;
+                category = OTHER_LABEL;
+            }
+            DynamicLabelType::KeyOnly(_) => {
+                if key_count >= MAX_LABELS || !label_is_valid(key, glean, meta) {
+                    key = OTHER_LABEL;
+                }
+            }
+            DynamicLabelType::CategoryOnly(_) => {
+                if category_count >= MAX_LABELS || !label_is_valid(category, glean, meta) {
+                    category = OTHER_LABEL;
+                }
+            }
+            DynamicLabelType::KeyAndCategory(_) => {
+                if key_count >= MAX_LABELS || !label_is_valid(key, glean, meta) {
+                    key = OTHER_LABEL;
+                }
+                if category_count >= MAX_LABELS || !label_is_valid(category, glean, meta) {
+                    category = OTHER_LABEL;
+                }
+            }
+        }
+        combine_base_identifier_and_labels(base_identifier, key, category)
+    } else {
+        record_error(
+            glean,
+            meta,
+            ErrorType::InvalidLabel,
+            "Invalid `DualLabeledCounter` label format, unable to determine key and/or category",
+            None,
+        );
+        combine_base_identifier_and_labels(base_identifier, OTHER_LABEL, OTHER_LABEL)
+    }
+}
+
+fn label_is_valid(label: &str, glean: &Glean, meta: &CommonMetricDataInternal) -> bool {
+    if label.len() > MAX_LABEL_LENGTH {
+        let msg = format!(
+            "label length {} exceeds maximum of {}",
+            label.len(),
+            MAX_LABEL_LENGTH
+        );
+        record_error(glean, meta, ErrorType::InvalidLabel, msg, None);
+        false
+    } else if label.chars().any(|c| !c.is_ascii() || c.is_ascii_control()) {
+        let msg = format!("label must be printable ascii, got '{:?}'", label);
+        record_error(glean, meta, ErrorType::InvalidLabel, msg, None);
+        false
+    } else {
+        true
+    }
+}
+
+fn count_keys_and_categories(meta: &CommonMetricDataInternal, glean: &Glean) -> (usize, usize) {
+    let base_identifier = &meta.base_identifier();
+    let prefix = format!("{base_identifier}{RECORD_SEPARATOR}");
+    let mut seen_keys: Vec<String> = Vec::new();
+    let mut seen_categorires: Vec<String> = Vec::new();
+    let mut snapshotter = |mut metric_id: &[u8], _: &Metric| {
+        let mut metric_id_str = String::new();
+        _ = metric_id.read_to_string(&mut metric_id_str);
+
+        // Split full identifier on the ASCII Record Separator (\x1e)
+        let parts: Vec<&str> = metric_id_str.split(RECORD_SEPARATOR).collect();
+
+        if parts.len() == 2 {
+            if !seen_keys.contains(&parts[0].into()) {
+                seen_keys.push(parts[0].into());
+            }
```

Comment:
Use a `HashSet` for `seen_keys` and `seen_categories` instead. Then all you need is `insert`.